### PR TITLE
chore: port cached channel pool to veneer backend

### PR DESCRIPTION
**Pull request by igorbernstein2**
This is an existing feature in bigtable-client-core that needed to be re-implemented for the transition to java-bigtable

### CODE REVIEW

**Comment by igorbernstein2** [comment]
Tests are fixed in #2917

**Review by mutianf — CHANGES_REQUESTED** [review]
Looks good, just noticed a typo in the class name :), and should the key use credentials instead of credentialsProvider?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/SharedDataClientWrapperFactory.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+   * Identity for a shared {@link ClientContext}.
+   *
+   * <p>This value class contains primary identifying information for a ClientContext. It can be
+   * extracted from {@link BigtableDataSettings} and be used to check if a ClientContext would be
+   * compatible with a ClientContext required by {@link BigtableDataSettings}.
+   */
+  static final class Key {
+    private final String endpoint;
+    private final Map<String, String> headers;
+    private final CredentialsProvider credentialsProvider;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/SharedDataClientWrapperFactory.java:110]: It has to be credentialsProvider. This class is used to check if we have a ClientContext for the given StubSettings. StubSettings contain providers, which will in most instances return a new instance of Credentials when getCredentials() is called. So the intention here is to reuse the same ClientContext when the StubSettings are using the same provider

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/SharedDataClentWrapper.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import com.google.protobuf.ByteString;
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+
+/** Simple wrapper around a {@link DataClientWrapper} that will release resources on close. */
+class SharedDataClentWrapper implements DataClientWrapper {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/SharedDataClentWrapper.java:38]: 😳

**Review by mutianf — APPROVED** [review]

**Pull request merged**